unity/unity-embed-host/CoreCLRHost.cs

```diff
@@ -41,15 +56,109 @@ internal static void InitState()
     public static IntPtr /*Assembly*/ load_assembly_from_data(byte* data, long size)
     {
         var assembly = alcWrapper.CallLoadFromAssemblyData(data, size);
-        return (IntPtr)assemblyHandleField.GetValue(assembly);
+        var ptr = GetHandleForAssembly(assembly);
+        return ptr;
     }
 
     [NativeFunction(NativeFunctionOptions.DoNotGenerate)]
     public static IntPtr /*Assembly*/ load_assembly_from_path(byte* path, int length)
     {
         var assembly = alcWrapper.CallLoadFromAssemblyPath(Encoding.UTF8.GetString(path, length));
-        return (IntPtr)assemblyHandleField.GetValue(assembly);
+        var ptr = GetHandleForAssembly(assembly);
+        return ptr;
+    }
+
+    private static IntPtr GetHandleForAssembly(Assembly assembly)
+    {
+        if (!m_assemblies.TryGetValue(assembly, out var pair))
```

Comment:
This is still racey.  Two threads can make it past the `if (!m_assemblies.TryGetValue(assembly, out var pair))`.  Each will then allocate a GC handle.  And one thread will "win" and have it's value stored in the dictionary.  There won't be a crash, but I believe for our use case we do not want 2 different GC handles to be floating around for the same assembly.

Reply:
I would suggest just locking now for simplicity/clarity perhaps.